File: api/src/modules/auth/auth-user.service.ts
import { Injectable } from "@nestjs/common";

import { DatabaseService } from "../../common/database/database.service";
import { assertUuid } from "../../common/utils/uuid";
import { AppRole } from "./interfaces/authenticated-user.interface";

@Injectable()
export class AuthUserService {
  constructor(private readonly databaseService: DatabaseService) { }

  async syncUserFromToken(
    userId: string,
    roles: AppRole[],
    publicUserId?: string
  ): Promise<void> {
    assertUuid(userId, "token.sub");
    const role = this.resolvePrimaryRole(roles);
    const normalizedPublicUserId = this.normalizePublicUserId(publicUserId, userId);

    await this.databaseService.query(
      `
      INSERT INTO users (id, role, username)
      VALUES ($1::uuid, $2::user_role, $3::text)
      ON CONFLICT (id)
      DO UPDATE SET
        role = EXCLUDED.role,
        username = COALESCE(NULLIF(EXCLUDED.username, ''), users.username),
        updated_at = now()
      `,
      [userId, role, normalizedPublicUserId]
    );
  }

  async getUsernameByUserId(userId: string): Promise<string | null> {
    assertUuid(userId, "userId");
    const result = await this.databaseService.query<{ username: string | null }>(
      `SELECT username FROM users WHERE id = $1::uuid`,
      [userId]
    );
    return result.rows[0]?.username ?? null;
  }

  private resolvePrimaryRole(roles: AppRole[]): AppRole | "both" {
    if (roles.includes("admin")) {
      return "admin";
    }
    if (roles.includes("support")) {
      return "support";
    }
    return "both";
  }

  private normalizePublicUserId(value: string | undefined, userId: string): string {
    const raw = (value ?? "").trim().toLowerCase();
    if (raw.length >= 3 && raw.length <= 40 && /^[a-z0-9._-]+$/.test(raw)) {
      return raw;
    }
    return `member_${userId.replace(/-/g, "").slice(0, 10).toLowerCase()}`;
  }
}

File: api-java/src/main/java/com/illamhelp/api/auth/AuthUserService.java
package com.illamhelp.api.auth;

import java.util.List;
import java.util.Optional;
import java.util.UUID;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class AuthUserService {
  private final UserRepository userRepository;

  public AuthUserService(UserRepository userRepository) {
    this.userRepository = userRepository;
  }

  @Transactional
  public void syncUserFromToken(String userId, List<String> roles, String publicUserId) {
    String role = roles.contains("admin") ? "admin" : roles.contains("support") ? "support" : "both";
    String username = normalizePublicUserId(publicUserId, userId);
    userRepository.upsertFromToken(userId, role, username);
  }

  public Optional<String> getUsernameByUserId(String userId) {
    return userRepository.findById(UUID.fromString(userId)).map(UserEntity::getUsername);
  }

  private String normalizePublicUserId(String value, String userId) {
    String raw = value == null ? "" : value.trim().toLowerCase();
    if (raw.length() >= 3 && raw.length() <= 40 && raw.matches("^[a-z0-9._-]+$")) {
      return raw;
    }
    return "member_" + userId.replace("-", "").substring(0, Math.min(10, userId.length()));
  }
}
